uc-crux-llvm/src/UCCrux/LLVM/Classify.hs

```diff
@@ -0,0 +1,546 @@
+                                       "(" <> Text.pack (show (LLVMPtr.ppPtr ptr)) <> ")"
+                                     ]
+                               unfixable appCtx tag
+                   _ -> unclass appCtx badBehavior
+             LLVMErrors.BBUndefinedBehavior
+               (UB.MemsetInvalidRegion (Crucible.RV ptr) _fillByte (Crucible.RV len)) ->
+                 case (getPtrOffsetAnn ptr, What4.asConcrete len) of
+                   ( Just
+                       ( Some
+                           ( TypedSelector
+                               ftRepr
+                               (SomeInSelector (SelectArgument idx cursor))
+                             )
+                         ),
+                     Just concreteLen
+                     ) ->
+                       do
+                         let tag = ArgMemsetTooSmall
+                         liftIO $
+                           (appCtx ^. log) Hi $
+                             Text.unwords
+                               [ "Diagnosis:",
+                                 diagnose tag,
+                                 "#" <> Text.pack (show (Ctx.indexVal idx)),
+                                 "at",
+                                 Text.pack (show (ppCursor (argName idx) cursor)),
+                                 "(" <> Text.pack (show (LLVMPtr.ppPtr ptr)) <> ")"
+                               ]
+                         liftIO $ (appCtx ^. log) Hi $ prescribe tag
+                         case ftRepr of
+                           FTPtrRepr partTypeRepr ->
+                             return $
+                               ExMissingPreconditions
+                                 ( tag,
+                                   oneArgShapeConstraint
+                                     idx
+                                     cursor
+                                     ( Initialized
+                                         (elemsFromOffset' concreteLen partTypeRepr)
+                                     )
+                                 )
+                           _ -> panic "classify" ["Expected pointer type"]
+                   _ -> unclass appCtx badBehavior
+             LLVMErrors.BBUndefinedBehavior
+               (UB.PoisonValueCreated poison) ->
+                 classifyPoison appCtx funCtx sym annotations poison
+                   >>= \case
+                     Nothing -> unclass appCtx badBehavior
+                     Just expl -> pure expl
+             LLVMErrors.BBMemoryError
+               ( MemError.MemoryError
+                   (summarizeOp -> (_expl, ptr))
+                   MemError.UnwritableRegion
+                 ) ->
+                 case getPtrOffsetAnn ptr of
+                   Just (Some (TypedSelector ftRepr (SomeInSelector (SelectArgument idx cursor)))) ->
+                     -- TODO: Double check that it really was unmapped not read-only
+                     -- or something?
```

Comment:
Well, I was adding some tests to make sure I was asking for review on working code, and what do you know, I uncovered a bug :sweat_smile: The program added in 8af76031e51ee98ad05318cc1aa2578daad92488 will make this heuristic start an infinite loop.

Reply:
To fix this will require a little bit of doing: We have to have access to the arguments that were passed to the argument to notice that the argument in question was already allocated in the first place. Not too bad though, I should have it done this afternoon. It'll also enable other kinds of heuristics.

Reply:
Fixed!